fourier.c

```diff
@@ -45,42 +45,65 @@
 #define VERSION "0.4.5"
 #endif
 
+#if (GIMP_MAJOR_VERSION == 3) || ((GIMP_MAJOR_VERSION == 2) && (GIMP_MINOR_VERSION >= 99))
+#ifdef HAVE_GETTEXT
+#include <libintl.h>
+#define d_(String) String
+#ifdef gettext_noop
+#    define N_(String) gettext_noop (String)
+#else
+#    define N_(String) (String)
+#endif
+#define _(String) gettext (String)
+#else
+/* No i18n for now */
+#define N_(x) x
+#define d_(x) x
+#define _(x) x
+#endif
+#else
+/* no gettext used with gimp2 version */
+#define N_(x) x
+#define d_(x) x
+#define _(x) x
+#endif
+
 /** Defines ******************************************************************/
 
 #define PLUG_IN_BINARY "fourier"
 #define PLUG_IN_NAME "plug_in_fft"
 #define PLUG_IN_VERSION "Jun 2024, " VERSION
 
-#define PLUG_IN_AUTHOR "Remi Peyronnet"
-#define PLUG_IN_MENU_LOCATION "<Image>/Filters/Generic"
+static char *PLUG_IN_AUTHOR = "Remi Peyronnet";
+static char *PLUG_IN_MENU_LOCATION = d_("<Image>/Filters/Generic");
 
-#define PLUG_IN_PROC "plug-in-fourier"
+static char *PLUG_IN_PROC = d_("plug-in-fourier");
 
-#define PLUG_IN_DIR_PROC "plug-in-fourier-forward"
-#define PLUG_IN_DIR_MENU_LABEL "FFT Forward"
-#define PLUG_IN_DIR_SHORT_DESC "This plug-in applies a FFT to the image, for educationnal or effects purpose."
-#define PLUG_IN_DIR_DESC  "Apply an FFT to the image. This can remove (for example) moire patterns from images scanned from books:\n\n" \
-                          "    The image should be RGB (Image|Mode|RGB)\n\n" \
-                          "    Remove the alpha layer, if present (Image|Flatten Image)\n\n" \
-                          "    Select Filters|Generic|FFT Forward\n\n" \
-                          "    Use the preselected neutral grey to effectively remove any moir patterns from the image. Either paint over any patterns or\n\n" \
-                          "     - In the Layers window, select the layer, and 'Duplicate Layer'\n" \
-                          "     - Select Colours|Brightness-Contrast. Increase the Contrast to see any patterns.\n" \
-                          "     - Use the Rectangular and/or Elliptical Selection tools to select any patterns on the contrast layer.\n" \
-                          "     - Then remove the contrast layer leaving the original FFT layer with the selections.\n" \
-                          "     - Then select Edit|Fill with FG colour, remembering to cancel the Selection afterwards!\n\n" \
-                          "    Select Filters|Generic|FFT Inverse\n\n" \
-                          "Voila, an image without the moire pattern!"
+static char *PLUG_IN_DIR_PROC = d_("plug-in-fourier-forward");
+static char *PLUG_IN_DIR_MENU_LABEL = d_("FFT Forward");
+static char *PLUG_IN_DIR_SHORT_DESC = d_("This plug-in applies a FFT to the image, for educational or effects purpose.");
+static char *PLUG_IN_DIR_DESC = d_("Apply an FFT to the image. This can remove (for example) moire patterns from images scanned from books:\n\n" \
```

Comment:
What is the reason to convert the #define in static char *?

I guess this is what causes the need of d_?

Reply:
```
#define PLUG_IN_DIR_PROC = "string"
gettext(PLUG_IN_DIR_PROC);
```
In this case, each reference to the defined PLUG_IN_DIR_PROC creates a string, so the program can look like this:
```
#define PLUG_IN_DIR_PROC = "string"
gettext("string_1");
gettext("string_2");
gettext("string_3");
gettext("string_4");
```
...and we also don't seem to have a method to xgettext or gettext the info.

Doing it like this, we can re-use the same string over and over again:
```
static char *ptr = "string";
gettext(ptr);
gettext(ptr);
gettext(ptr);
gettext(ptr);
```
...and we also have a method to xgettext the string for the po file, while gettext("string") translates the string.

xgettext is looking for strings that contain 'd_()', while gettext is matching po file "string" with the contents of *ptr which is "string", and then giving you a locale version translation.

Reply:
Seems to me a little complicated to maintain to gain a few kb (you need to keep in sync the use of d_ and _), but OK

Reply:
Today's PCs have abundant memory, less than $1 per gigabyte, so memory is indeed cheap.
The only benefits you can consider are you're shaving a few bytes download for the distros running on a shoestring budget, and the number of CPU cycles shaved in loading a smaller binary per user. LOL - saving the planet from byte bloat.
The main target of this exercise was to get xgettext to find strings to add to gimp30-fourier.pot, since you can't add *ptrs or defines in the *.po files, and in this way, it served it's purpose - xgettext finds the strings :-)

Reply:
FYI I have refactored in the 2.99 branch a bit the gettext setup part to be more compliant at the way GIMP has designed the use of i18n with the set_i18n callback function.

It has the advantage to handle paths relative to the plugin path to add i18n also for win32 build. 

I have overridden this default path to keep the one you used for Linux packaging, but as I am not able for now to test Linux builds, I hope that have not broken things

Reply:
I removed existing fourier plugin, and tried again including your patches, plus my added patches, seems to work okay.
I also did a 'make dist' and with the tar.gz file tested 'rpmbuild -ba' and that worked okay too.
I added some of your improvements to gimp3-fix-ca, but for some reason I just wasn't able to make use of your po/Makefile.am line:
```
$(GETTEXT_PACKAGE3).pot: ../fourier.c
```
...(using ../fix-ca) which didn't work for my po/Makefile.am. I even tried {top_srcdir)/fix-ca.c but that didn't work either even though this area of the files are fairly similar. Puzzled.

Reply:
Found a problem. even though 'make install' installs both languages, it seems 'rpmbuild -ba' only installs fr.mo, but not pt.mo.
With gimp3-fix-ca it works ok and installs all languages. Seems LANGUAGES only contains "fr", and not "fr pt" when running `rpmbuild -ba`

Reply:
I reverted the faulty `$(GETTEXT_PACKAGE3).pot: ../fourier.c`  ; I noticed the pot was not refreshed automatically so it was an attempt to solve that, but if it causes problems manual run should be enough.

I have tried to see what we could have missed with pt not installed with rpmbuild but I have found nothing... Have you tried with a fresh clone?

Reply:
We had the problem that the pot and po files were constantly getting updated in fontforge, thus giving a false impression that the files were up to date, when in fact, some files were years out of date.
Based on experience, I'd advise to let the po files remain static (not updated).
The pot file, unsure what to say here, maybe update it when you intend to update 1 or more po files? or leave instructions on what to do if you intend to make or update a po file, as described in po/Makefile.am

Reply:
The rpm pt.* files is a little bit operator error on my part.
Your project is called plugin-gimp-fourier, and the projects in fedora follow the naming scheme gimp-fourier-plugin.
With other projects, I would normally do `autoconf -i;automake;./configure;make dist` and then copy the gimp-fourier-plugin-0.4.5.tar.gz into ~/rpmbuild/SOURCES/ and then run `rpmbuild -ba ~/rpmbuild/SPECS/gimp3-fourier-plugin.spec` file, and watch the results.
Apparently, I was testing over-n-over again the ~/rpmbuild/SOURCES/plugin-gimp-fourier-0.4.5.gz file made back in April, so it didn't contain pt.* files of course.
I deleted the old files, today and tried building again, by synthesizing a fresh plugin-gimp-fourier-0.4.5.tar.gz file. I found 1 bug and fixed it now for gimp2, will shortly try gimp3 ....

Reply:
Appears the fix was smaller than expected, seems to work okay running `./configure --enable-gimp3;make dist` as per #17